pkg/kademlia/lookup.go

```diff
@@ -38,60 +44,116 @@ func newSequentialLookup(rt *RoutingTable, nodes []*pb.Node, client node.Client,
-			return ctx.Err()
-		default:
-		}
+func (lookup *lookup) Run(ctx context.Context) error {
+	wg := sync.WaitGroup{}
+	nowork := sync.Cond{L: lookup.rMutex}
+	working := 0
+	allDone := false
+
+	wg.Add(lookup.routines)
+	for i := 0; i < lookup.routines; i++ {
+		go func(i int) {
+			defer func() {
+				wg.Done()
+			}()
+			for {
+				var item *Item
+
+				// protect the state with a lock
+				lookup.rMutex.Lock()
+				for {
+					// when everything is done then we can just leave
+					if allDone {
+						lookup.rMutex.Unlock()
+						return
+					}
+
+					// try to get work
+					var ok bool
+					item, ok = lookup.queue.Pop().(*Item)
+					if ok {
+						// mark ourselves as working
+						working++
+						break
+					}
+
+					// no work, let's wait until some other worker pushes new work
+					nowork.Wait()
+				}
+				lookup.rMutex.Unlock()
+
+				// do the work and find new work items
+				targetBytes := new(big.Int).SetBytes(lookup.target.Bytes())
+				additionalWork := lookup.work(ctx, item, targetBytes)
+
+				// protect the state
+				lookup.rMutex.Lock()
+
+				// add things to the queue
+				for _, work := range additionalWork {
+					lookup.queue.Push(work)
+					// wake up one worker
+					nowork.Signal()
+				}
+
+				// say that we are done
+				working--
+
+				// the last worker marks all done
+				allDone = working == 0 && len(lookup.queue) == 0
+
+				lookup.rMutex.Unlock()
+
+				// tell waiting routines to check allDone
+				nowork.Broadcast()
+			}
+		}(i)
+	}
 
-		item := heap.Pop(&lookup.queue).(*Item)
-		if !lookup.bootstrap && item.priority.Cmp(zero) == 0 {
-			// found the result
-			return nil
-		}
-		next := item.value
+	wg.Wait()
+	// TODO(bryanchriswhite) is this correct?
+	return ctx.Err()
+}
 
-		neighbors := lookup.FetchNeighbors(ctx, next)
-		for _, neighbor := range neighbors {
-			if lookup.contacted[neighbor.GetId()] {
-				continue
-			}
+func (lookup *lookup) work(ctx context.Context, item *Item, targetBytes *big.Int) []*Item {
+	var newWork = []*Item{}
+	next := item.value
 
-			priority := new(big.Int).Xor(targetBytes, new(big.Int).SetBytes(lookup.target.Bytes()))
-			heap.Push(&lookup.queue, &Item{
-				value:    neighbor,
-				priority: priority,
-			})
+	neighbors := lookup.FetchNeighbors(ctx, next)
+	for _, neighbor := range neighbors {
+		if lookup.contacted[neighbor.GetId()] {
+			continue
 		}
 
-		for len(lookup.queue) > lookup.limit {
-			heap.Pop(&lookup.queue)
-		}
+		priority := new(big.Int).Xor(targetBytes, new(big.Int).SetBytes(lookup.target.Bytes()))
+		newWork = append(newWork, &Item{
+			value:    neighbor,
+			priority: priority,
+		})
 	}
-	return nil
+
+	return newWork
 }
 
-func (lookup *sequentialLookup) FetchNeighbors(ctx context.Context, node *pb.Node) []*pb.Node {
+func (lookup *lookup) FetchNeighbors(ctx context.Context, node *pb.Node) []*pb.Node {
 	if node.GetAddress() == nil {
 		return nil
 	}
+	lookup.lMutex.Lock()
 	lookup.contacted[node.GetId()] = true
```

Comment:
nit -- interesting thing I just learned from @stefanbenten is it's more efficient to do `map[string]{}`

https://emersion.fr/blog/2017/sets-in-go/